File: apps/backend/scripts/check_access_control.go
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/ethclient"

	"apps/backend/contracts/eventaccessmanager"
)

func main() {
	rpcURL := os.Getenv("BLOCKCHAIN_RPC_URL")
	if rpcURL == "" {
		rpcURL = "https://eth-sepolia.g.alchemy.com/v2/p-1v26f8urtKJAgsL3ry5"
	}

	client, err := ethclient.Dial(rpcURL)
	if err != nil {
		log.Fatal("Failed to connect to blockchain:", err)
	}

	eventAccessManagerAddr := common.HexToAddress("0x0EADEa85ee88907c46A0EF6AF2f0C530D55BEBB6")
	recoveredSigner := common.HexToAddress("0x9F6f8ef7c3CD068e96C6643754dbF57A49ee13aB")
	systemTransactor := common.HexToAddress("0xf466e7cE6B06f9b3071557A790Bd45F051C1C60A")

	fmt.Println("🔍 Checking Access Control...")
	fmt.Printf("EventAccessManager: %s\n", eventAccessManagerAddr.Hex())
	fmt.Printf("Recovered Signer: %s\n", recoveredSigner.Hex())
	fmt.Printf("System Transactor: %s\n\n", systemTransactor.Hex())

	accessManager, err := eventaccessmanager.NewEventAccessManager(eventAccessManagerAddr, client)
	if err != nil {
		log.Fatal("Failed to instantiate EventAccessManager:", err)
	}

	// Check 1: Get DECM_ACCESS_MANAGER address
	fmt.Println("━━━ CHECK 1: DECM_ACCESS_MANAGER Address ━━━")
	decmAccessManagerAddr, err := accessManager.DECMACCESSMANAGER(nil)
	if err != nil {
		fmt.Printf("❌ Error getting DECM_ACCESS_MANAGER: %v\n", err)
	} else {
		fmt.Printf("DECM_ACCESS_MANAGER: %s\n", decmAccessManagerAddr.Hex())
		if decmAccessManagerAddr == (common.Address{}) {
			fmt.Println("⚠️  WARNING: DECM_ACCESS_MANAGER is zero address!")
		}
	}

	// Check 2: Is recovered signer a host?
	fmt.Println("\n━━━ CHECK 2: Is Recovered Signer a Host? ━━━")
	isHost, err := accessManager.CheckIsHost(nil, recoveredSigner)
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
	} else {
		if isHost {
			fmt.Println("✅ YES - Recovered signer IS a host")
		} else {
			fmt.Println("❌ NO - Recovered signer is NOT a host")
		}
	}

	// Check 3: Is recovered signer host or admin?
	fmt.Println("\n━━━ CHECK 3: Is Recovered Signer Host/Admin? ━━━")
	isHostOrAdmin, err := accessManager.CheckIsHostOrAdmin(nil, recoveredSigner)
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
	} else {
		if isHostOrAdmin {
			fmt.Println("✅ YES - Recovered signer IS host or admin")
		} else {
			fmt.Println("❌ NO - Recovered signer is NOT host or admin")
		}
	}

	// Check 4: Is system transactor allowed?
	fmt.Println("\n━━━ CHECK 4: Is System Transactor Allowed? ━━━")
	isAllowed, err := accessManager.CheckIsAllowedMsgSender(nil, systemTransactor)
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
	} else {
		if isAllowed {
			fmt.Println("✅ YES - System transactor IS allowed")
		} else {
			fmt.Println("❌ NO - System transactor is NOT allowed")
		}
	}

	fmt.Println("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	fmt.Println("VERDICT:")
	fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
	if isHostOrAdmin || isAllowed {
		fmt.Println("✅ Access control is satisfied")
		if isHostOrAdmin {
			fmt.Println("   - Recovered signer is host/admin")
		}
		if isAllowed {
			fmt.Println("   - System transactor is allowed")
		}
	} else {
		fmt.Println("❌ Access control is NOT satisfied")
		fmt.Println("   - Recovered signer is NOT host/admin")
		fmt.Println("   - System transactor is NOT allowed")
		fmt.Println("\n💡 FIX NEEDED:")
		fmt.Printf("   1. Add %s as host in EventAccessManager\n", recoveredSigner.Hex())
		fmt.Printf("   2. OR add %s to allowed senders in DecmAccessManager\n", systemTransactor.Hex())
	}
}

File: scripts/check-access-control.ts
#!/usr/bin/env tsx

import { ethers } from "ethers";

// Configuration - FROM YOUR LOGS
const RPC_URL = process.env.BLOCKCHAIN_RPC_URL || "http://localhost:8545";
const RECOVERED_SIGNER = "0x9F6f8ef7c3CD068e96C6643754dbF57A49ee13aB";
const SYSTEM_TRANSACTOR = "0xf466e7cE6B06f9b3071557A790Bd45F051C1C60A";

// !!! REPLACE THIS WITH THE FULL 42-CHARACTER ADDRESS FROM DATABASE !!!
const CERTIFICATE_CONTRACT = "0x3B328a7049a374a5D728d575383d38486f1c727c";

// Certificate Contract ABI (minimal - to get EventAccessManager address)
const CERTIFICATE_ABI = ["function EVENT_ACCESS_MANAGER() external view returns (address)"];

// EventAccessManager ABI (minimal)
const ACCESS_MANAGER_ABI = [
    "function checkIsHostOrAdmin(address account) external view returns (bool)",
    "function checkIsAllowedMsgSender(address account) external view returns (bool)",
    "function getHosts() external view returns (address[])",
    "function getAllowedMsgSenders() external view returns (address[])",
];

async function checkAccessControl() {
    console.log("🔍 Checking Access Control...\n");

    const provider = new ethers.JsonRpcProvider(RPC_URL);

    // First, get EventAccessManager address from certificate contract
    console.log("📡 Fetching EventAccessManager address from certificate contract...");
    const certificateContract = new ethers.Contract(
        CERTIFICATE_CONTRACT,
        CERTIFICATE_ABI,
        provider,
    );

    const EVENT_ACCESS_MANAGER_ADDRESS = await certificateContract.EVENT_ACCESS_MANAGER();
    console.log(`✅ Found EventAccessManager at: ${EVENT_ACCESS_MANAGER_ADDRESS}\n`);

    const accessManager = new ethers.Contract(
        EVENT_ACCESS_MANAGER_ADDRESS,
        ACCESS_MANAGER_ABI,
        provider,
    );

    console.log("Configuration:");
    console.log(`  RPC: ${RPC_URL}`);
    console.log(`  Certificate Contract: ${CERTIFICATE_CONTRACT}`);
    console.log(`  EventAccessManager: ${EVENT_ACCESS_MANAGER_ADDRESS}`);
    console.log(`  Recovered Signer: ${RECOVERED_SIGNER}`);
    console.log(`  System Transactor: ${SYSTEM_TRANSACTOR}\n`);

    // Check 1: Is recovered signer a host/admin?
    console.log("━━━ CHECK 1: Is Recovered Signer a Host/Admin? ━━━");
    try {
        const isHost = await accessManager.checkIsHostOrAdmin(RECOVERED_SIGNER);
        console.log(`Result: ${isHost ? "✅ YES" : "❌ NO"}`);

        if (!isHost) {
            console.log("\n⚠️  Recovered signer is NOT registered as host/admin!");
            console.log("This is likely the issue.\n");
        }
    } catch (error) {
        console.error("❌ Error checking host status:", error);
    }

    // Check 2: Is system transactor allowed?
    console.log("\n━━━ CHECK 2: Is System Transactor Allowed? ━━━");
    try {
        const isAllowed = await accessManager.checkIsAllowedMsgSender(SYSTEM_TRANSACTOR);
        console.log(`Result: ${isAllowed ? "✅ YES" : "❌ NO"}`);

        if (!isAllowed) {
            console.log("\n⚠️  System transactor is NOT in allowed senders list!");
            console.log("This could be the issue if recovered signer is also not host/admin.\n");
        }
    } catch (error) {
        console.error("❌ Error checking allowed sender:", error);
    }

    // Check 3: List all hosts
    console.log("\n━━━ CHECK 3: All Registered Hosts ━━━");
    try {
        const hosts = await accessManager.getHosts();
        console.log(`Total hosts: ${hosts.length}`);
        hosts.forEach((host: string, i: number) => {
            const match = host.toLowerCase() === RECOVERED_SIGNER.toLowerCase();
            console.log(`  ${i + 1}. ${host} ${match ? "← MATCH!" : ""}`);
        });
    } catch (error) {
        console.error("❌ Error listing hosts:", error);
    }

    // Check 4: List all allowed senders
    console.log("\n━━━ CHECK 4: All Allowed Senders ━━━");
    try {
        const senders = await accessManager.getAllowedMsgSenders();
        console.log(`Total allowed senders: ${senders.length}`);
        senders.forEach((sender: string, i: number) => {
            const match = sender.toLowerCase() === SYSTEM_TRANSACTOR.toLowerCase();
            console.log(`  ${i + 1}. ${sender} ${match ? "← MATCH!" : ""}`);
        });
    } catch (error) {
        console.error("❌ Error listing allowed senders:", error);
    }

    // Final verdict
    console.log("\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    console.log("VERDICT:");
    console.log("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\n");

    try {
        const isHost = await accessManager.checkIsHostOrAdmin(RECOVERED_SIGNER);
        const isAllowed = await accessManager.checkIsAllowedMsgSender(SYSTEM_TRANSACTOR);

        if (!isHost && !isAllowed) {
            console.log("❌ NEITHER condition is satisfied!");
            console.log("   - Recovered signer is NOT host/admin");
            console.log("   - System transactor is NOT allowed sender");
            console.log("\n💡 FIX: Add one of the following:");
            console.log(`   1. Register ${RECOVERED_SIGNER} as host in EventAccessManager`);
            console.log(`   2. Add ${SYSTEM_TRANSACTOR} to allowed senders in EventAccessManager`);
        } else if (isHost && isAllowed) {
            console.log("✅ BOTH conditions are satisfied - this should work!");
            console.log("   The issue might be something else.");
        } else if (isHost) {
            console.log("✅ Recovered signer IS host/admin - should work!");
        } else if (isAllowed) {
            console.log("✅ System transactor IS allowed sender - should work!");
        }
    } catch (error) {
        console.error("❌ Error determining verdict:", error);
    }
}

checkAccessControl()
    .then(() => process.exit(0))
    .catch((error) => {
        console.error("Fatal error:", error);
        process.exit(1);
    });
